python/python/tests/test_pubsub.py

```diff
@@ -0,0 +1,486 @@
+        CHANNEL_NAMES = {"test-channel-1", "test-channel-2"}
+        MESSAGES_BY_CHANNEL = {
+            channel : {
+                id : {
+                    channel: {
+                        "on_" + channel + "_" + msg + str(id): 1 for msg in ["foo_from_", "bar_from_", "baz_from_"]
+                    }
+                } for id in range(CLIENTS_COUNT_PER_CHANNEL)
+            } for channel in CHANNEL_NAMES
+        }
+
+        all_clients: List[TestPubSub.PubSubClient] = []
+        for channel in CHANNEL_NAMES:
+            for id in range(CLIENTS_COUNT_PER_CHANNEL):
+                all_clients.append(TestPubSub.PubSubClient(
+                    subsciptions={
+                        CoreCommands.ChannelModes.Exact: {channel}
+                    },
+                    sharded_messages_to_publish=MESSAGES_BY_CHANNEL[channel][id] if sharded_pubsub else [],
+                    unsharded_messages_to_publish=MESSAGES_BY_CHANNEL[channel][id] if not sharded_pubsub else [],
+                    cluster_mode=cluster_mode,
+                    request=request))
+                all_clients[-1].connect_and_subscribe()
+        
+        TestPubSub.publish_and_assert(all_clients)
+
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    async def test_pubsub_channel_boundary_pattern(self, request, cluster_mode):
+        """ Tests that messages do not cross channel boundaries by pattern subsctiption """
+
+        CLIENTS_COUNT_PER_CHANNEL = 5
+        CHANNEL_NAMES = {"test-channel-1", "test-channel-2"}
+        CHANNEL_PATTERNS = {"test?channel-1", "test?channel-2"}
+        MESSAGES_BY_CHANNEL = {
+            channel : {
+                id : {
+                    channel: {
+                        "on_" + channel + "_" + msg + str(id): 1 for msg in ["foo_from_", "bar_from_", "baz_from_"]
+                    }
+                } for id in range(CLIENTS_COUNT_PER_CHANNEL)
+            } for channel in CHANNEL_NAMES
+        }
+
+        all_clients: List[TestPubSub.PubSubClient] = []
+        for channel_id in len(CHANNEL_NAMES):
+            for id in range(CLIENTS_COUNT_PER_CHANNEL):
+                all_clients.append(TestPubSub.PubSubClient(
+                    subsciptions={
+                        CoreCommands.ChannelModes.Pattern: {CHANNEL_PATTERNS[channel_id]}
+                    },
+                    sharded_messages_to_publish=[],
+                    unsharded_messages_to_publish=MESSAGES_BY_CHANNEL[CHANNEL_NAMES[channel_id]][id],
+                    cluster_mode=cluster_mode,
+                    request=request))
+                all_clients[-1].connect_and_subscribe()
+        
+        TestPubSub.publish_and_assert(all_clients)
+
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    @pytest.mark.parametrize("sharded_pubsub", [True, False])
+    async def test_pubsub_callbacks_per_channel(self, request, cluster_mode, sharded_pubsub):
+        """ Tests that callbacks are per channel """
```

Comment:
you should also check that the correct context is passed

Reply:
done